src/turnkeyml/analyze/discover.py

```diff
@@ -0,0 +1,246 @@
+import argparse
+import copy
+import os
+import inspect
+from typing import Optional, List
+import torch
+import turnkeyml.build.stage as stage
+import turnkeyml.common.exceptions as exp
+import turnkeyml.common.filesystem as fs
+from turnkeyml.analyze.script import (
+    evaluate_script,
+    TracerArgs,
+)
+import turnkeyml.common.printing as printing
+
+
+default_max_depth = 0
+
+
+class Discover(stage.Stage):
+    """
+    Discover the PyTorch models and their corresponding inputs in a python script (.py)
+    and send one model/inputs pair onwards into the sequence.
+
+    Expected inputs: None
+
+    Outputs:
+        state.results: a PyTorch model instance (torch.nn.Module)
+        state.inputs: a dictionary of example inputs to the model's forward function,
+            e.g., model(**inputs)
+    """
+
+    unique_name = "discover"
+
+    def __init__(self):
+        super().__init__(monitor_message="Discovering PyTorch models")
+
+    @staticmethod
+    def parser(add_help: bool = True) -> argparse.ArgumentParser:
+        parser = argparse.ArgumentParser(
+            description="Discover the PyTorch models in a python script",
+            add_help=add_help,
+        )
+
+        parser.add_argument(
+            "--script-args",
+            dest="script_args",
+            type=str,
+            help="Arguments to pass into the target script(s)",
+        )
+
+        parser.add_argument(
+            "--max-depth",
+            dest="max_depth",
+            type=int,
+            default=default_max_depth,
+            help="Maximum depth to analyze within the model structure of the target script(s)",
+        )
+
+        # Hidden argument required by TurnkeyML for any stage that starts a sequence
+        parser.add_argument("--input", help=argparse.SUPPRESS)
+
+        return parser
+
+    def parse(self, state: fs.State, args, known_only=True) -> argparse.Namespace:
+        parsed_args = super().parse(state, args, known_only)
+
+        file_path, targets, encoded_input = fs.decode_input_arg(parsed_args.input)
+
+        parsed_args.input = file_path
+
+        if len(targets) > 1:
+            raise exp.ArgError(
+                "Only one target (number after the ::) is allowed, "
+                f"but received {encoded_input}"
+            )
+        elif len(targets) == 1:
+            parsed_args.target = targets[0]
+        else:  # len(targets)==0
```

Comment:
```suggestion
        else:
```

Reply:
Wont fix: i find it helpful to label else statements if they have a specific intent. that way if the if or elif statements change, you know if you need to change the else to sustain the intent.